sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala

```diff
@@ -736,15 +691,156 @@ private[execution] final class UnboundedFollowingWindowFunctionFrame(
 
     // Only recalculate and update when the buffer changes.
     if (bufferUpdated) {
-      evaluatePrepared(buffer, inputIndex, buffer.length)
-      fill(target, outputIndex)
+      processor.initialize(input.size)
+      processor.update(input, inputIndex, input.size)
+      processor.evaluate(target)
     }
 
     // Move to the next row.
     outputIndex += 1
   }
+}
+
+/**
+ * This class prepares and manages the processing of a number of aggregate functions.
+ *
+ * This implementation only supports evaluation in [[Complete]] mode. This is enough for
+ * Window processing.
+ *
+ * Processing of distinct aggregates is currently not supported.
+ *
+ * The implementation is split into an object which takes care of construction, and a the actual
+ * processor class. Construction might be expensive and could be separated into a 'driver' and a
+ * 'executor' part.
+ */
+private[execution] object AggregateProcessor {
+  def apply(functions: Array[Expression],
+    ordinal: Int,
+    inputAttributes: Seq[Attribute],
+    newMutableProjection: (Seq[Expression], Seq[Attribute]) => () => MutableProjection):
+    AggregateProcessor = {
+    val aggBufferAttributes = mutable.Buffer.empty[AttributeReference]
+    val initialValues = mutable.Buffer.empty[Expression]
+    val updateExpressions = mutable.Buffer.empty[Expression]
+    val evaluateExpressions = mutable.Buffer.fill[Expression](ordinal)(NoOp)
+    val imperatives = mutable.Buffer.empty[ImperativeAggregate]
+
+    // Create and add a size reference to SizeBasedWindowFunction.
+    var sizeOrdinal = -1
+    var size: BoundReference = null
+    val addSize = (f: Expression) => f match {
+      case wf: SizeBasedWindowFunction =>
+        if (size == null) {
+          sizeOrdinal = aggBufferAttributes.size
+          size = BoundReference(sizeOrdinal, IntegerType, false)
+          aggBufferAttributes += wf.size
+          initialValues += NoOp
+          updateExpressions += NoOp
+        }
+        wf.withSize(size)
```

Comment:
We - unfortunately - do ```wf.size``` is ref 'owned' by one of the ```SizeBasedWindowFunction```s, all other size expressions have different ```size``` properties, so they won't bind to this one.

I'll try to see if we can make this work in a more transparant way, by sharing the ```size``` propery (or its id).